aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java

```diff
@@ -168,56 +169,42 @@ public Map<String, String> properties() {
   @Override
   public void deleteFiles(Iterable<String> paths) throws BulkDeletionFailureException {
     if (awsProperties.s3DeleteTags() != null && !awsProperties.s3DeleteTags().isEmpty()) {
+      Set<Tag> deleteTags = awsProperties.s3DeleteTags();
       Tasks.foreach(paths)
           .noRetry()
           .executeWith(executorService())
           .suppressFailureWhenFinished()
           .onFailure(
-              (path, exc) ->
-                  LOG.warn(
-                      "Failed to add delete tags: {} to {}",
-                      awsProperties.s3DeleteTags(),
-                      path,
-                      exc))
-          .run(path -> tagFileToDelete(path, awsProperties.s3DeleteTags()));
+              (path, exc) -> LOG.warn("Failed to add delete tags: {} to {}", deleteTags, path, exc))
+          .run(path -> tagFileToDelete(path, deleteTags));
     }
 
-    if (!awsProperties.isS3DeleteEnabled()) {
-      return;
+    if (awsProperties.isS3DeleteEnabled()) {
+      SetMultimap<String, String> bucketToObjects = computeBucketToObjects(paths);
```

Comment:
We are now eagerly computing the bucket to objects mapping up front, as opposed to before where we would iterate over the paths, keep track of the objects per bucket and if for a given bucket the size of the objects is the batch size, the deletion would get triggered (and the mapping would get removed). 

Now, it's all up front, so there would be more memory consumption but I think it should be ok. 1 million objects with a max key of 1024 bytes is 1 GB representation of paths.

Reply:
1gb does sounds like a lot , example on the spark driver.  Why cant we do it the old way?

Reply:
can we try to use https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/s3/paginators/ListObjectsV2Iterable.html so that the list is dynamically loaded instead of buffered upfront

Reply:
@jackye1995 I don't think we need a separate API actually, we are already given the list of paths. We can still do it lazily by basically using the previous approach and constructing the bucket/key from the path and keeping track of when the objects for a bucket hits a certain batch size. It's just that using the Tasks framework in its current state becomes infeasible since it operates on a list and also will wait for completion. We want to just submit the batch deletion and move on, and then at the very end check the statuses of all those tasks. So using the underlying executor service fits that pattern. Let me know what you think.

Reply:
Sounds good!